monty.qmd

```diff
@@ -1,16 +1,95 @@
 
-* defining a model in monty, both easily with `monty_function` and less easily with `monty_model`
-* fitting models with MCMC
-* writing priors with the `monty_dsl`
-* using observers to follow along with the model (if we can come up with a good example)
-* working with nested models (once we write the code that would allow this)
-* running MCMC chains in parallel
-* exporting chains to work with other packages
-* different samplers
+1. Simple R code,
+2. A dedicated domain-specific language (DSL),
+3. Integration of `odin` models,
+4. Composition of two `monty` models.
 
-Hmm, there really is quite a bit of ground to cover here!
+The most basic method to define a `monty` model is through a simple R function, as shown in this chapter. More advanced examples are covered in the @sec-monty-models in particular in the context of Bayesian statistics. @sec-monty-dsl introduces the monty DSL for more versatile model building using a small probabilistic DSL similar to BUGs. @sec-monty-combine explains how to compose new monty models by combining two existing ones. The last part of this book starting from @sec-inference demonstrates how to incorporate `odin` models into `monty` workflows.
+
+## A simple example
+
+```{r}
+library(monty)
+```
+
+We can define a simple Gaussian [mixture model](https://en.wikipedia.org/wiki/Mixture_model) of two sub-populations using the `monty_model_function()` from the package. The model represents the distribution of a quantity $l$, sampled with a probability $p$ from a normal distribution of mean $m_{1}$ and with a probability $1-p$ from another normal distribution of mean $m_{2}$. Both subpopulations have variance equal to 1. 
+
+To build our `monty` model, we start by defining an R function returning the log-density of our statistical model. This function has for arguments $l$, the quantity of interest and the three parameters $p$, $m_{1}$ and $m_{1}$ defining the two Normally-distributed subpopulations.
+
+```{r}
+fn <- function(l, p, m1, m2) {
+  log(p*dnorm(l,mean = m1) + (1-p)*dnorm(l, mean = m2))
+}
+```
+
+Assuming that the population is 75% from subpopulation 1 (normally distributed with mean 3) and 25% from subpopulation 2 (also normally distributed but with mean 7), we can build our `monty` model by indicating that the parameters of the subpopulations are fixed at these given values.
+
+```{r}
+l_distribution <- monty_model_function(fn,
+                                       fixed = list(p = 0.75, m1 = 3, m2=7))
+```
+
+We have just created a `monty` model. 
+
+```{r}
+l_distribution
+```
+
+We can plot the density of our model for values of $l$ between 0 and 10 and eye-check that we can see the two subpopulations in the correct proportions and with the 3 and 7 as modes.
+
+```{r}
+l <- seq(from = 0, to = 10, by = 0.1)
+plot(l,
+     exp(Vectorize(l_distribution$density)(l)),
```

Comment:
@richfitz I need to use Vectorize() here to pass multiple value of the argument - let me know if a better interface can be shown.

Reply:
See above. You will need to pass a 1-row matrix in to vectorise , so

```r
l <- matrix(seq(from = 0, to = 10, by = 0.1), 1)
monty_distribution_density(l_distribution, l)
```

Please do not use `$density` in docs

Reply:
Not sure how to make this work. The following code:
`fn <- function(l, p, m1, m2) {
  log(p * dnorm(l, mean = m1) + (1 - p) * dnorm(l, mean = m2))
}
l_distribution <- monty_model_function(fn,
                                       fixed = list(p = 0.75, m1 = 3, m2 = 7),
                                       allow_multiple_parameters = TRUE)
l <- matrix(seq(from = 0, to = 10, by = 0.1), 1)
monty_model_density(l_distribution, l)`

Return the following error, that I interpret as a conflict with having some parameters set as fixed. I could pass a 101x4 instead with the fixed argument like this (that works fine),

`l_distribution <- monty_model_function(fn,
                                       allow_multiple_parameters = TRUE)

l <- matrix(c(seq(from = 0, to = 10, by = 0.1), rep(0.75,101), rep(3,101), rep(7,101)),4)
monty_model_density(l_distribution, l)`

but then it makes things quite more complex to understand and it's not the same as the monty model as then 4 "parameters" with a mixed interpretation instead of one.

`Error in `packer$unpack()`:
! Can't unpack a matrix where the unpacker uses 'fixed'
Run `rlang::last_trace()` to see where the error occurred.`